src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -316,7 +316,10 @@ internal bool TryWrite(Utf8JsonWriter writer, in T value, JsonSerializerOptions
             }
 
             bool ignoreCyclesPopReference = false;
+
             if (options.ReferenceHandlingStrategy == ReferenceHandlingStrategy.IgnoreCycles &&
+                // .NET types that are serialized as JSON primitive values don't need to be tracked for cycle detection e.g: string.
+                ConverterStrategy != ConverterStrategy.Value &&
```

Comment:
How does this work with custom (user) converters (also `ConverterStrategy.Value` and the instance could be a non-primitive)? Do we need to also check `IsInternalConverter`?

Reply:
Based on local testing, I think we should not push to the stack when the type has a custom converter under the principle that the user should take care of serialization him/herself and because `Preserve` doesn't preserve objects with custom converters.

Reply:
>Do we need to also check `IsInternalConverter`?

Doesn't `ConverterStrategy != ConverterStrategy.Value` suffice for guarding against custom (user) converters?

Reply:
FWIW: Prior to this change, objects that used an internal converter were failing in the same fashion as the string case reported in the linked issue. 
I added a couple of test cases to cover said scenario (https://github.com/dotnet/runtime/pull/52253/commits/a97c9ea8d8175a0c55e0ab91c4481c71c752d011).